### DISCUSSION TITLE
Make Void as a valid return type

### DISCUSSION THREAD

**Discussion by marius02** [dotnet/csharplang#4616] [Language Ideas] [answered]
Right now System.Void is not a valid return type in C#.

``` 
            public async Task<Unit> Handle(Command request, CancellationToken cancellationToken)
            {
                var activity = await _context.Activities.FindAsync(request.Activity.Id);
                _mapper.Map(request.Activity, activity);

                await _context.SaveChangesAsync(cancellationToken);
                return Unit.Value;
            }
```

In this case Unit just simulates a void return.

Are there any restrictios, why Void cannot be a valid return type?

**Comment by HaloFour** [comment, marked as answer]
`System.Void` is not permitted to be used as a generic type argument by the runtime and by the ECMA 335 specification.

See: https://github.com/dotnet/csharplang/discussions/1604
https://github.com/dotnet/csharplang/discussions/696
https://github.com/dotnet/csharplang/discussions/2767
https://github.com/dotnet/csharplang/discussions/1279
https://github.com/dotnet/csharplang/issues/3288

**Comment by juwens** [comment]
I see this as a problem too.

People need to implement workarounds like this in every advanced project: https://chtenb.dev/?page=unit-cs

I guess adding a Unit type to the standard library would be nice. (like they did in F#)

**Comment by tacosontitan** [comment]
Is there a reason you can't just use `Task` as the return type?

```cs
public async Task Handle(Command request, CancellationToken cancellationToken)
{
    var activity = await _context.Activities.FindAsync(request.Activity.Id);
    _mapper.Map(request.Activity, activity);
    await _context.SaveChangesAsync(cancellationToken);
}
```

**Reply by leo-costa** [reply]
I assume because you need to create the same function for anything returning Task<T>. You duplicate the function because you can't return void.
